libcnb/src/layer/handling.rs

```diff
@@ -262,7 +265,17 @@ fn write_layer<M: Serialize, P: AsRef<Path>>(
                 fs::create_dir_all(&exec_d_dir)?;
 
                 for (name, path) in exec_d_programs {
-                    fs::copy(path, exec_d_dir.join(name))?;
+                    // We could just try to copy the file here and let the call-site deal with the
+                    // IO errors when the path does not exist. We're using an explicit error variant
+                    // for a missing exec.d binary makes it easier to debug issues with packaging
+                    // since the usage of exec.d binaries often relies on implicit packaging the
+                    // buildpack author might not be aware of.
+                    Some(&path)
+                        .filter(|path| path.exists())
+                        .ok_or_else(|| WriteLayerError::MissingExecDFile(path.clone()))
+                        .and_then(|path| {
+                            fs::copy(path, exec_d_dir.join(name)).map_err(WriteLayerError::IoError)
+                        })?;
```

Comment:
Implementing using a conditional rather than a combinator seems a bit simpler, since:
- it avoids the need to wrap in `Some()`
- means the `map_err` is no longer required
- means the `.clone()` is no longer required

```suggestion
                    if !path.exists() {
                        return Err(WriteLayerError::MissingExecDFile(path));
                    }
                    fs::copy(path, exec_d_dir.join(name))?;
```

However I'm fine with it either way.